File: SemanticSearch/Java/Web/SpringBoot/src/main/java/com/portfolio/semanticsearch/vectorstore/InMemoryVectorStore.java
package com.portfolio.semanticsearch.vectorstore;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class InMemoryVectorStore implements VectorStoreAdapter {
    private final Map<String, CollectionData> collections = new LinkedHashMap<>();

    private record CollectionData(List<String> docs, List<float[]> embeddings, List<Map<String, Object>> metadata) {
    }

    @Override
    public void connect() {
    }

    @Override
    public void addDocuments(List<String> documents, List<float[]> embeddings, List<Map<String, Object>> metadata) {
        String collection = System.getenv("VECTOR_COLLECTION") != null ? System.getenv("VECTOR_COLLECTION") : "documents";
        CollectionData data = collections.computeIfAbsent(collection, k -> new CollectionData(new ArrayList<>(), new ArrayList<>(), new ArrayList<>()));
        data.docs.addAll(documents);
        data.embeddings.addAll(embeddings);
        data.metadata.addAll(metadata);
    }

    @Override
    public List<SearchResult> search(float[] queryEmbedding, int nResults) {
        String collection = System.getenv("VECTOR_COLLECTION") != null ? System.getenv("VECTOR_COLLECTION") : "documents";
        CollectionData data = collections.get(collection);
        if (data == null || data.docs.isEmpty()) {
            return new ArrayList<>();
        }
        List<SearchResult> results = new ArrayList<>();
        int count = Math.min(nResults, data.docs.size());
        for (int i = 0; i < count; i++) {
            results.add(new SearchResult(data.docs.get(i), data.metadata.get(i), 0.0));
        }
        return results;
    }

    @Override
    public void deleteCollection(String collectionName) {
        collections.remove(collectionName);
    }

    @Override
    public List<String> listCollections() {
        return new ArrayList<>(collections.keySet());
    }

    @Override
    public void close() {
        collections.clear();
    }
}

File: SemanticSearch/Elixir/Web/Phoenix/src/lib/semantic_search/vectorstore/in_memory_vector_store.ex
defmodule SemanticSearch.VectorStore.InMemoryVectorStore do
  @moduledoc """
  In-memory vector store: stores docs/embeddings/metadata per collection in an
  Agent. Used as the default/fallback store and in tests.
  """
  use Agent

  @behaviour SemanticSearch.VectorStore.VectorStoreAdapter

  alias SemanticSearch.VectorStore.SearchResult

  def start_link(opts \\ []) do
    Agent.start_link(fn -> %{} end, opts ++ [name: __MODULE__])
  end

  @impl true
  def connect, do: :ok

  @impl true
  def add_documents(documents, embeddings, metadata) do
    collection = collection_name()

    Agent.update(__MODULE__, fn state ->
      data =
        Map.get(state, collection, %{documents: [], embeddings: [], metadata: []})

      data = %{
        documents: data.documents ++ documents,
        embeddings: data.embeddings ++ embeddings,
        metadata: data.metadata ++ metadata
      }

      Map.put(state, collection, data)
    end)

    :ok
  end

  @impl true
  def search(_query_embedding, n_results) do
    collection = collection_name()

    data = Agent.get(__MODULE__, &Map.get(&1, collection))

    case data do
      nil ->
        []

      %{documents: docs, metadata: _metas} when docs == [] ->
        []

      %{documents: docs, metadata: metas} ->
        docs
        |> Enum.zip(metas)
        |> Enum.take(n_results)
        |> Enum.map(fn {doc, meta} ->
          %SearchResult{document: doc, metadata: meta, distance: 0.0}
        end)
    end
  end

  @impl true
  def delete_collection(name) do
    Agent.update(__MODULE__, &Map.delete(&1, name))
    :ok
  end

  @impl true
  def list_collections do
    Agent.get(__MODULE__, &Map.keys/1)
  end

  defp collection_name, do: System.get_env("VECTOR_COLLECTION", "documents")
end
